homeassistant/components/transmission.py

```diff
@@ -0,0 +1,193 @@
+"""
+Component for monitoring the Transmission BitTorrent client API.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/transmission/
+"""
+from datetime import timedelta
+
+import logging
+import voluptuous as vol
+
+from homeassistant.util import Throttle
+from homeassistant.helpers import discovery
+from homeassistant.helpers.event import track_time_interval
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_NAME,
+    CONF_PORT,
+    CONF_USERNAME,
+    CONF_MONITORED_CONDITIONS,
+)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['transmissionrpc==0.11']
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'transmission'
+DATA_TRANSMISSION = 'data_transmission'
+
+DEFAULT_NAME = 'Transmission'
+DEFAULT_PORT = 9091
+TURTLE_MODE = 'turtle_mode'
+
+SENSOR_TYPES = {
+    'active_torrents': ['Active Torrents', None],
+    'current_status': ['Status', None],
+    'download_speed': ['Down Speed', 'MB/s'],
+    'paused_torrents': ['Paused Torrents', None],
+    'total_torrents': ['Total Torrents', None],
+    'upload_speed': ['Up Speed', 'MB/s'],
+    'completed_torrents': ['Completed Torrents', None],
+    'started_torrents': ['Started Torrents', None],
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(TURTLE_MODE, default=False): cv.boolean,
+        vol.Optional(CONF_MONITORED_CONDITIONS, default=['current_status']):
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SCAN_INTERVAL = timedelta(minutes=2)
+
+
+def setup(hass, config):
+    """Set up the Transmission Component."""
+    host = config[DOMAIN][CONF_HOST]
+    username = config[DOMAIN][CONF_USERNAME]
+    password = config[DOMAIN][CONF_PASSWORD]
+    port = config[DOMAIN][CONF_PORT]
+
+    import transmissionrpc
+    from transmissionrpc.error import TransmissionError
+    try:
+        api = transmissionrpc.Client(
+            host, port=port, user=username, password=password)
+        api.session_stats()
+    except TransmissionError as error:
+        if str(error).find("401: Unauthorized"):
```

Comment:
Is it ok to continue after all other errors caught here?

It would be best to make the distinction in the library, by subclassing a base library specific error.

Reply:
This question hasn't been answered.

Reply:
What do you mean with "continue after all other errors"?

If you mean that TransmissionError is a too generic kind of error, and so it may catch other kind of error, you're right. But in any error case the setup of the platform should stop. Since I'm not the owner of the transmissionrpc package I can't make modification to implement other kind of more specific error.

Reply:
I'm assuming there are other error types since you're checking for a string in the error argument, and only upon match we treat it as something to act on.

Checking for a string seems fragile. But I guess it's ok, if there's no better way right now.

Reply:
I see now that it's only the log message that's indented under the if check.

Reply:
Yeah I was about to comment on that. It’s the previous behavior, I agree that it should be improved in order to log the error message in any case... maybe in another PR :D